dsgrid/query/models.py

```diff
@@ -0,0 +1,305 @@
+        return check_aggregations(aggregations)
+
+    @validator("output_format")
+    def check_format(cls, fmt):
+        allowed = {"csv", "parquet"}
+        if fmt not in allowed:
+            raise ValueError(f"output_format={fmt} is not supported. Allowed={allowed}")
+        return fmt
+
+
+class ProjectQueryResultModel(QueryResultBaseModel):
+    """Represents a user query on a Project."""
+
+    project: ProjectConstraintsModel = Field(
+        title="project", description="Defines the datasets to use and how to transform them."
+    )
+
+    @root_validator
+    def check_metric_reductions(cls, values):
+        if values["metric_reductions"] and values["project"].metric_reductions:
+            raise ValueError(
+                "metric_reductions cannot be set within the project constraints as well as in the result"
+            )
+        return values
+
+    def serialize_cached_content(self):
+        # Exclude all result-oriented fields in orer to faciliate re-using queries.
+        text = self.project.json(indent=2)
+        return compute_hash(text.encode()), text
+
+
+class CreateDerivedDatasetQueryModel(QueryBaseModel):
+    """Represents a user query to create a Derived Dataset. This dataset requires a Project
+    in order to retrieve dimension records and dimension mapping records.
+    """
+
+    dataset_id: str = Field(title="dataset_id", description="Derived Dataset ID for query")
+    project: ProjectConstraintsModel = Field(
+        title="project", description="Defines the datasets to use and how to transform them."
+    )
+    supplemental_columns: List[str] = Field(
+        title="supplemental_columns",
+        description="Add these supplemental dimension query names as columns in result tables. "
+        "Applies to all aggregations.",
+        default=[],
+    )
+    metric_reductions: List[MetricReductionModel] = Field(
+        title="metric_reductions",
+        description="Specifies how metric values should be reduced.",
+        default=[],
+    )
+    aggregations: List[Union[AggregationModel, ChainedAggregationModel]] = Field(
+        title="aggregations", description="Informs how to groupBy and aggregate data.", default=[]
+    )
+
+    @validator("aggregations")
+    def check_aggregations(cls, aggregations):
+        return check_aggregations(aggregations)
+
+    def serialize_cached_content(self):
+        # Exclude all result-oriented fields in orer to faciliate re-using queries.
+        text = self.project.json(indent=2)
+        return compute_hash(text.encode()), text
+
+
+# class CreateStandaloneDerivedDatasetQueryModel(QueryBaseModel):
+# """Represents a user query to create a Standalone Derived Dataset. This dataset contains
+# all dimension records and dimension mapping records. It does not need a Project.
+# """
+# TODO: create the format (config / .toml) that will define this object
+
+
+class DerivedDatasetQueryModel(QueryResultBaseModel):
```

Comment:
Are there equivalencies between these terms? If so, I think we should rename for consistency. Which one is the standalone dataset query interface (i.e. without a dependency on project)?

- ProjectConstraintModel ~ CreateDerivedDatasetQueryModel
- ProjectQueryResultModel ~ DerivedDatasetQueryModel

Reply:
ProjectConstraintsModel is used for both project queries (ProjectQueryResultModel) and derived-dataset-creation queries (CreateDerivedDatasetQueryModel). It defines the constraints that are used to filter the data.

The output of ProjectQueryResultModel cannot be queried; it is a final result.
The output of CreateDerivedDatasetQueryModel can be queried - with DerivedDatasetQueryModel. That query produces a final result.

I would like to have a group discussion on these workflows to ensure they are ideal. They work in some cases but probably won't in all. Then we'll have a better idea on naming conventions.

Reply:
DerivedDatasetQueryModel should include the word `Result`